tracer/src/Datadog.Trace/DatabaseMonitoring/DatabaseMonitoringPropagator.cs

```diff
@@ -73,5 +75,79 @@ internal static string PropagateSpanData(DbmPropagationLevel propagationStyle, s
 
             return string.Empty;
         }
+
+        /// <summary>
+        /// Uses a sql instruction to set a context for the current connection, bearing the span ID and trace ID.
+        /// This is meant to circumvent cache invalidation issues that occur when those values are injected in comment.
+        /// Currently only working for MSSQL (uses an instruction that is specific to it)
+        /// </summary>
+        /// <returns>True if the traceparent information was set</returns>
+        internal static bool PropagateDataViaContext(Tracer tracer, DbmPropagationLevel propagationLevel, IntegrationId integrationId, IDbConnection? connection, string serviceName, Scope scope, SqlTags tags)
+        {
+            if (propagationLevel != DbmPropagationLevel.Full || integrationId != IntegrationId.SqlClient || connection == null)
+            {
+                return false;
+            }
+
+            // we want the instrumentation span to be a sibling of the actual query span
+            var instrumentationParent = scope.Parent?.Span?.Context;
+            // copy the tags so that modifications on one span don't impact the other
+            var copyProcessor = new ITags.CopyProcessor<SqlTags>();
+            tags.EnumerateTags(ref copyProcessor);
+            using (var instrumentationScope = tracer.StartActiveInternal("set context_info", instrumentationParent, tags: copyProcessor.TagsCopy, serviceName: serviceName))
```

Comment:
I'm concerned that adding an additional span that is not part the users normal trace will lead to noisy traces. It also increases payload size considerably, which users are billed for in APM.

Did we consider keeping this span behind a configuration setting? I would argue we should keep it disabled it by default, and allow enabling it for troubleshooting purposes. But if we want it enabled by default, we should at least allow users to disable it if they feel it's too noisy.

Reply:
well, it is kinda behind a config setting, because "full" mode is off by default, but you mean still having full mode without the span.
I agree with you, it was pretty hard to get an opinion from product on this, and in the Java implem we ended up putting the span.
It essentially doubles the number of sql spans when enabled, which can have a lot of impact indeed.

Since it's a relatively new feature, we can still change the default for Java (or have different defaults depending on languages).
The alternative we thought of was to just add a tag on the query span with the time it took to do the "set context", just so that the info is available.

Reply:
> it is kinda behind a config setting, because "full" mode is off by default

Yeah, maybe this may be enough for now.

Reply:
There's another issue to consider: the sqlclient service will now have _two_ top-level operation names (`sql-server.query` and the new one), which leads to a diminishes the UX. Datadog will pick one operation name as the default and show a drop-down menu in the UI for users to switch between them.

And in the edge case where the sqlclient span was the root span, the new span added here will be in a _separate_ span from the original span.

I think that making the new span a child (instead of a sibling) of the original span would solve both of these issues.

Reply:
ah good point... Can a child be in the past ? The idea why we wanted to have this a sibling rather than a child is that we didn't want to affect the duration of the span covering the actual query to not distort stats on span length whether this is enabled or not

Reply:
> Can a child be in the past ?

It probably works, but I'm not completely certain. It should be easy enough to test, though.

> we didn't want to affect the duration of the span

From the user's pov, we _are_ making the query take longer, so maybe we should include this? I could see it go either way 😅

Another alternative could be to add a new parent span for the two siblings. Benefits include: we show the original query time, we show the "combined" query time (including the `set context` query), and there's a parent span to keep the two sibling together. Another benefits, from a trace stats pov, is that there will be only one service-entry span instead of two.

Something like:

```
+---------------------------------+
| NEW PARENT SPAN                 |
+---------------------------------+
+------------++ +-----------------+
| SET CONTEXT | | SQL QUERY       |
+------------++ +-----------------+

```

Reply:
hmm, yeah but that's tripling the number of spans generated by the sql integration instead of doubling, and also we have to find a meaningful name for yet an other span 😅

Reply:
> that's tripling the number of spans generated

Yes, I agree that's an issue. But I see too many issues with adding a sibling span:
- it creates two traces if there is no parent span
- it creates two service-entry spans, which I think can create issues with trace stats
```
[ set context ] [ sql query ]
```

I'm not sure what the best solution is, but I am leaning towards:
- _not_ adding the new span (avoids the whole problem 😅 )
- adding the new span as a child of the original span (and including it in the original span's duration)
```
[ --------------- sql query ]
 [ set context ]
```
- adding the new span as a child of the original span (and _not_ including it in the original span's duration)
```
                [ sql query ]
[ set context ]
```

Reply:
This may need more design discussion including folks from the APM Astra team and such.